### PR TITLE
implement IPersistTimeoutsV2 interface

### PR DESCRIPTION
**Pull request by timbussmann**
Implement the IPersistTimeoutsV2 interface in order to fix loosing timeout messages when not using dtc.

see issue https://github.com/Particular/NServiceBus/issues/2885

Connects to Particular/NServiceBus#2885

### CODE REVIEW

**Comment by timbussmann** [comment]
@Particular/nservicebus nhibernate experts please review

**Comment by ramonsmits** [comment]
As discussed during the review:

* Normal operation is peek, send, delete. During peek we can do a row lock so that the delete will always succeed. Currently the peek and delete are executed in the same transaction when using connection sharing between sql transport and nhibernate persistence. Maybe this can potentially happen when running with multiple instances are fetching timeouts as these instance are competing on the timeout queue. (`ISession.Get<T>(object, LockMode.Upgrade`)
* The entity mapping is duplicate except for one property. Unknown why this is or what behavioral change this will have. Maybe refactor to one mapping due to the number of properties and using object initializer instead of a constructor which can potentially result in missing properties if the object is extended in the future.

**Comment by timbussmann** [comment]
@ramonsmits just had a look at the `LockMode.Upgrade` when peeking but I get some unwanted behavior in the unit test. If you have two concurrent transactions creating an update lock, can they deadlock each other?

**Comment by ramonsmits** [comment]
No, that should not be the case as yo do a SELECT FOR UPDATE, that should not result in a deadlock, it can only result in a sequential operation. 2nd transaction will have to wait before the first completes but the 1st does not have to wait for the 2nd because it already acquired a row write lock.

**Comment by timbussmann** [comment]
@ramonsmits never mind, It seems to be an issue with SQLite, same code works on SQLExpress.

**Comment by timbussmann** [comment]
@ramonsmits please review the latest commit, I changed the code according to your points:
* using upgrade lock for peek. This should only improve scenarios when using dtc for now, since as we saw, this version doesn't yet support connection sharing. So the only case where that lock may be held until the TryRemove is for ambient transactions.
* extracted the entity mapping so both `Peek` and `TryRemove` behave the same.

File: src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_fetching_timeouts_from_storage.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ namespace NServiceBus.TimeoutPersisters.NHibernate.Tests
     public class When_fetching_timeouts_from_storage : InMemoryDBFixture
     {
         [Test]
-        public void Should_return_the_complete_list_of_timeouts()
+        public void GetNextChunk_should_return_the_complete_list_of_timeouts()
```
Comment (ramonsmits) [inline src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_fetching_timeouts_from_storage.cs]: Why the method name change? Test is unaltered?

File: src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs [hunk 2]
```diff
@@ -79,30 +81,97 @@ public void Should_remove_timeouts_by_id()
...
+
+            Assert.IsTrue(t1Result.HasValue && t2Result.HasValue);
+
+            // one delete should succeed, the other one shouldn't
+            Assert.IsTrue(t1Result.Value || t2Result.Value);
+            Assert.IsFalse(t1Result.Value && t2Result.Value);
+        }
+
+        [Test]
+        public void RemiveTimeoutBy_should_remove_timeouts_by_sagaid()
```
Comment (ramonsmits) [inline src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs]: Typo in method

File: src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs [hunk 3]
```diff
@@ -79,30 +81,97 @@ public void Should_remove_timeouts_by_id()
...
 
         [Test]
-        public void Should_remove_timeouts_by_sagaid()
+        public void TryRemove_should_return_false_when_timeout_already_deleted()
+        {
+            var timeout = new TimeoutData();
+
+            persister.Add(timeout);
+
+            Assert.IsTrue(persister.TryRemove(timeout.Id));
```
Comment (ramonsmits) [inline src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs]: How is this even possible? The timeout being stored has no ID assigned. Is it assigned by the database?

File: src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs [hunk 4]
```diff
@@ -79,30 +81,97 @@ public void Should_remove_timeouts_by_id()
...
+            Assert.IsFalse(persister.TryRemove(timeout.Id));
+        }
+
+        [Test]
+        public void TryRemove_should_work_with_concurrent_transactions()
+        {
+            var timeout = new TimeoutData()
+            {
+                Id = Guid.NewGuid().ToString(),
```
Comment (ramonsmits) [inline src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs]: Conflicting behavior regarding ID assignment with test `TryRemove_should_return_false_when_timeout_already_deleted `

File: src/NServiceBus.NHibernate/TimeoutPersisters/TimeoutStorage.cs [hunk 5]
```diff
@@ -189,6 +180,47 @@ static string ConvertDictionaryToString(ICollection data)
             return serializer.SerializeObject(data);
         }
 
+        public TimeoutData Peek(string timeoutId)
```
Comment (ramonsmits) [inline src/NServiceBus.NHibernate/TimeoutPersisters/TimeoutStorage.cs]: As mentioned, I don't agree with this method name. Its an entity (get, delete) and not a message (peek/receive). Maybe an improvement for v6.

File: src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs]: yes, timeout id is always assigned by the timeout persister

File: src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_fetching_timeouts_from_storage.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_fetching_timeouts_from_storage.cs]: to distinct between `GetNextChunk` and `Peek` tests

File: src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs (hunk 4 above)
Comment (timbussmann) [inline src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs]: oh that id assignment shouldn't be there. Thats some remainder from the SQLite tests. Going to remove that.

File: src/NServiceBus.NHibernate/TimeoutPersisters/TimeoutStorage.cs (hunk 5 above)
Comment (timbussmann) [inline src/NServiceBus.NHibernate/TimeoutPersisters/TimeoutStorage.cs]: happy to discuss this on the v6 solution.

**Comment by timbussmann** [comment]
@Particular/nservicebus @ramonsmits any further concerns or can we merge this?

**Comment by andreasohlund** [comment]
Looks good to me

**Pull request merged**